### ISSUE TITLE
twisted.test.test_process.ProcessTestCase.testEcho times out intermittently

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3861]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3861|
|Type|defect|
|Created|2009-05-30 19:04:05Z|
|Branch|https://github.com/twisted/twisted/tree/slowecho-3861|

This especially seems to impact one Windows slave in particular:

http://buildbot.twistedmatrix.com/builders/winxp32-py2.4-select/builds/428/steps/select/logs/problems

<details><summary>Searchable metadata</summary>

```
trac-id__3861 3861
type__defect defect
reporter__glyph glyph
priority__low low
milestone__ 
branch__branches_slowecho_3861 branches/slowecho-3861
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1243710245000000 1243710245000000
changetime__1246355056000000 1246355056000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph removed owner|
|-|-|

Okay I have no idea how to fix this.  I've done some investigating, and gotten 200 subsequent successful test runs of the test case in question on a Windows machine.  I haven't tried python2.4 yet, but I doubt it would make too much of a difference.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

I got RDP access to the slave in question, and while I haven't actually gotten the test to _fail_, I have noticed that it is incredibly, excruciatingly slow.  Manually editing some files on the slave to remove the '-u' flag from the python subprocess seems to speed it up by at least `2000%` and not affect its semantics.

I think the test might be timing out just because unbuffered I/O is actually extremely slow.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

(the test is doing 70kb of I/0, one byte at a time)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [27003]) Branching to 'slowecho-3861'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

The branch has a simple hack in it now which should be suitable for running on the appropriate buildslave, when it comes back.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @glyph|
|-|-|

 *  would you mind adding a docstring to testEcho, and rename it test_echo?
 * test_unsetPid uses process_echoer as well, so "-u" can be removed.

Please merge once done and if you get a successful build. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Augh.  The buildslave is _finally_ back.  Two successful builds, one before addressing your concerns, one after.  Also I removed that bogus timeout which really shouldn't have been there anyway.  Landing shortly.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph set status to `closed`|
|-|-|

(In [27032]) Correct intermittent failure of process echo test.

This change makes `twisted.test.test_process.ProcessTestCase.test_echo` (nee `testEcho`, coding standard compliance also fixed by this change) run substantially faster by having its helper process explicilty flush its output rather than run in unbuffered mode.  In particular, this makes the test substantially faster on Windows, where it was so excruciatingly slow that it was causing the windows buildbot to fail with a timeout more often than not.

Author: glyph

Reviewer: therve

Fixes [#3861](https://github.com/twisted/twisted/issues/3861)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
